crates/runtime/src/datafusion/udf/bucket.rs

```diff
@@ -0,0 +1,464 @@
+/*
+Copyright 2024-2025 The Spice.ai OSS Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+     https://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+use std::num::TryFromIntError;
+use std::sync::{Arc, LazyLock};
+
+use ahash::RandomState;
+use arrow::array::{ArrayRef, UInt64Array};
+use arrow::compute::binary;
+use datafusion::arrow::array::{Array, Int32Array};
+use datafusion::arrow::datatypes::DataType;
+use datafusion::common::DataFusionError;
+use datafusion::common::hash_utils::create_hashes;
+use datafusion::logical_expr::{
+    ColumnarValue, ScalarFunctionArgs, ScalarUDFImpl, Signature, Volatility,
+};
+use datafusion::scalar::ScalarValue;
+use snafu::{ResultExt as _, Snafu};
+
+/// Maximum number of buckets, chosen to support large-scale partitioning while preventing excessive memory usage.
+const MAX_NUM_BUCKETS: i64 = 1_000_000;
+
+/// Fixed seed for deterministic hashing.
+const HASH_SEED: usize = 0x53_50_49_43_45;
```

Comment:
Can we also snapshot the value we get in the test? It wouldn't be good for this value to change across versions.